actions.go

```diff
@@ -3,10 +3,157 @@ package neo4j
 import (
 	"fmt"
 
+	"github.com/cloudprivacylabs/lsa/pkg/ls"
 	"github.com/cloudprivacylabs/opencypher/graph"
+	"github.com/fatih/structs"
 	"github.com/neo4j/neo4j-go-driver/v4/neo4j"
 )
 
+type neo4jAction interface {
+	Run(neo4j.Transaction) error
+}
+
+var neo4jActions = map[string]neo4jAction{
+	"Address":     recreateEntity{},
+	"Observation": insert{},
+}
+
+type deleteEntity struct {
+	Config
+	graph.Graph
+	entityId int64
+}
+
+type createEntity struct {
+	Config
+	graph.Graph
+	graph.Node
+}
+
+type recreateEntity struct {
+	ce createEntity
+	de deleteEntity
+}
+
+func (r recreateEntity) Run(tx neo4j.Transaction) error {
+	err := r.de.delete(tx)
+	if err != nil {
+		return err
+	}
+	err = r.ce.create(tx)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d deleteEntity) delete(tx neo4j.Transaction) error {
+	err := loadEntityNodes(tx, d.Graph, []int64{d.entityId}, d.Config, findNeighbors, nil)
+	if err != nil {
+		return err
+	}
+	// delete entity and all nodes reachable from this entity,
+	_, err = tx.Run("MATCH (m)-[*]->(n) WHERE ID(m)=$id DETACH DELETE m,n", map[string]interface{}{"id": d.entityId})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// used for query
+func mapNodes(nodes []neo4jNode) []map[string]interface{} {
+	var result = make([]map[string]interface{}, len(nodes))
+	for index, item := range nodes {
+		result[index] = structs.Map(item)
+	}
+	return result
+}
+
+func (c createEntity) create(tx neo4j.Transaction) error {
+	// TODO: recreate, find all reachable nodes from this node
+	// create entity
+	vars := make(map[string]interface{})
+	query := fmt.Sprintf("CREATE (m %s %s), return m",
+		c.MakeLabels(c.Node.GetLabels().Slice()),
+		c.MakeProperties(c.Node, vars))
+	idrec, err := tx.Run(query, vars)
+	if err != nil {
+		return err
+	}
+	rec, err := idrec.Single()
+	if err != nil {
+		return err
+	}
+	eid := rec.Values[0].(neo4j.Node).Id
+
+	connectedComponents := findConnectedComponents(c.Graph, c.Node, c.Config)
```

Comment:
Why is  this necessary? Given graph and a root node, you can already find all the nodes/edges in that entity.